### PR TITLE
Control GeoServer connections on URLs provided by clients/external systems

### PR DESCRIPTION
**Pull request by imranrajjad**
JIRA : https://osgeo-org.atlassian.net/browse/GEOT-6512
Will fail compile due to pre-requisite : https://github.com/geotools/geotools/pull/2804
## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.**

For all pull requests:

- [ ] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [ ] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [ ] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [ ] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [ ] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Commits changing the REST API, or any configuration object, should check if the REST API docs (Swagger YAML files and classic documentation) need to be updated.

### CODE REVIEW

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/RemoteRequestInputProvider.java [hunk 1]
```diff
@@ -60,7 +61,8 @@ public RemoteRequestInputProvider(
     protected Object getValueInternal(ProgressListener listener) throws Exception {
         InputReferenceType ref = input.getReference();
         URL destination = new URL(ref.getHref());
-
+        if(destination.getProtocol().toLowerCase().startsWith("http"))
```
Comment (aaime) [inline src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/RemoteRequestInputProvider.java:64]: Hum what about FTP? Is it even supported?

File: src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ExecuteTest.java [hunk 2]
```diff
@@ -751,6 +755,43 @@ public void testFeatureCollectionFileReferenceKVPError() throws Exception {
                 "/wps:ExecuteResponse/wps:Status/wps:ProcessFailed/ows:ExceptionReport/ows:Exception/ows:ExceptionText",
                 d);
     }
+    
+    @Test
+    public void testHttpSecutiry() throws Exception {
```
Comment (aaime) [inline src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ExecuteTest.java:760]:
```suggestion
    public void testHttpSecurity() throws Exception {
```

File: src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ExecuteTest.java [hunk 3]
```diff
@@ -751,6 +755,43 @@ public void testFeatureCollectionFileReferenceKVPError() throws Exception {
...
+        URLEntry googleOnly =
+                new URLEntry(
+                        "google only",
+                        "only allow url starting with http://www.google.com",
+                        "^(http://www.google).*$");
+        addURLEntryGeoserverURLConfigService(googleOnly);
+        //GeoserverURLConfigService.getSingleton().addAndsave(googleOnly);
```
Comment (aaime) [inline src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ExecuteTest.java:767]: Please do not leave commented out bits of code. Also, the formatter would have put a space between "//" and "GeoserverURLConfigService", so this did not go through a local maven build

File: src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ExecuteTest.java [hunk 4]
```diff
@@ -751,6 +755,43 @@ public void testFeatureCollectionFileReferenceKVPError() throws Exception {
...
+        Document d = getAsDOM(request);
+        // print(d);
+        checkValidationErrors(d);
+        
+        assertEquals("wps:ExecuteResponse", d.getDocumentElement().getNodeName());
+        
+        assertXpathExists("/wps:ExecuteResponse/wps:Status/wps:ProcessFailed", d);
+        assertXpathEvaluatesTo(
+                "Failed to retrieve value for input features\n" + 
+                "&gt;&gt;&gt;Evaluation Failure: http://www.unallowed.com/resource.xml: did not pass security evaluation",
```
Comment (aaime) [inline src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ExecuteTest.java:791]: I would say something like "Connections to http://www.unallowed.com/resource.xml are not allowed by the security configuration"

File: src/main/src/main/java/org/geoserver/catalog/HTTPStoreInfo.java [hunk 5]
```diff
@@ -65,4 +65,8 @@
      * @see #setMaxConnections(int)
      */
     public void setUseConnectionPooling(boolean useHttpConnectionPooling);
+
+    public boolean isUseSecuredHttp();
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/catalog/HTTPStoreInfo.java:69]: Sounds like HTTPS, does this imply using the URL verification instead? I would suggest to call the property "isUseURLVerificationCheks" (not necessarily HTTP). Please change in all places where "secured HTTP" is used, it's confusing.

File: src/main/src/main/java/org/geoserver/security/urlchecker/GeoserverURLChecker.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+    @Override
+    public boolean isEnabled() {
+        return enabled;
+    }
+
+    @Override
+    public boolean evaluate(String url) {
+        if (Dispatcher.REQUEST.get() == null) {
+            // ignore calls made from admin console
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/urlchecker/GeoserverURLChecker.java:65]:
```suggestion
            // Apply only to OGC requests
```

File: src/main/src/main/java/org/geoserver/security/urlchecker/GeoserverURLChecker.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+            return true;
+        }
+        if (url == null) return false;
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/urlchecker/GeoserverURLChecker.java:68]:
```suggestion
        if (url == null || url.isEmpty()) return false;

```

File: src/main/src/main/java/org/geoserver/security/urlchecker/GeoserverURLChecker.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        }
+
+        if (LOGGER.isLoggable(Level.FINE)) {
+            LOGGER.log(Level.FINE, url + " did not match any REGEX");
+        }
+
+        return false;
+    }
+
+    private List<URLEntry> getEnabled() {
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/urlchecker/GeoserverURLChecker.java:90]:
```suggestion
    private List<URLEntry> getEnabledUREntries() {
```

File: src/main/src/main/java/org/geoserver/security/urlchecker/GeoserverURLConfigService.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+    private static GeoserverURLConfigService singleton;
+
+    public GeoserverURLConfigService(
+            GeoServerResourceLoader dataDir, XStreamPersisterFactory xstreamPersisterFactory)
+            throws Exception {
+        xp = initialize(xstreamPersisterFactory.createXMLPersister());
+        Resource resource = dataDir.get("security//");
+        workspaceDirectory = resource.dir();
+        xmlFile = dataDir.find(workspaceDirectory, "urls.xml");
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/urlchecker/GeoserverURLConfigService.java:46]:
```suggestion
        xmlFile = dataDir.find(workspaceDirectory, "url-allow-list.xml");

```

File: src/main/src/main/java/org/geoserver/security/urlchecker/GeoserverURLConfigService.java [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+        return xstreamPersister;
+    }
+
+    private synchronized GeoserverURLChecker createFromScatch() {
+        ArrayList<URLEntry> list = new ArrayList<URLEntry>();
+        list.add(
+                new URLEntry(
+                        "generic",
+                        "allow http connections and disk access",
+                        "^(http|https|file)://.*$"));
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/urlchecker/GeoserverURLConfigService.java:83]: Maybe ftp too.

File: src/main/src/main/java/org/geoserver/security/urlchecker/GeoserverURLConfigService.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+    /**
+     * @return the geoserverURLChecker
+     * @throws Exception
+     */
+    public GeoserverURLChecker getGeoserverURLCheckerCopy() throws Exception {
+        return (GeoserverURLChecker) geoserverURLChecker.clone();
+    }
+
+    //    public static GeoserverURLConfigService getSingleton() {
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/urlchecker/GeoserverURLConfigService.java:162]: Please remove uncommented code.

File: src/main/src/main/java/org/geoserver/security/urlchecker/URLEntry.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
+/* (c) 2020 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.security.urlchecker;
+
+import java.io.Serializable;
+
+/** @author ImranR */
+public class URLEntry implements Serializable {
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/urlchecker/URLEntry.java:10]: Usual deal, please add javadoc to public classes and methods.

File: src/main/src/main/java/org/vfny/geoserver/util/Requests.java [hunk 13]
```diff
@@ -121,6 +122,9 @@ public static boolean isLoggedIn(HttpServletRequest request) {
      * @return The InputStream used to validate and parse the SLD xml.
      */
     public static InputStream getInputStream(URL url) throws IOException {
+        // handle security for http/https
+        if (url.getProtocol().toLowerCase().startsWith("http")) URLCheckers.evaluate(url);
```
Comment (aaime) [inline src/main/src/main/java/org/vfny/geoserver/util/Requests.java:126]: Do not limit to http (unless there is a reason that I'm missing?). E.g., one might now want the server to access the local file system out of params from remote request calls, seems a legit concern no?).

File: src/platform/src/test/java/org/geoserver/platform/resource/ResourcesTest.java [hunk 14]
```diff
@@ -8,8 +8,10 @@
 import static org.junit.Assert.assertTrue;
 
 import java.io.IOException;
+import org.junit.Ignore;
 import org.junit.Test;
 
+@Ignore
```
Comment (aaime) [inline src/platform/src/test/java/org/geoserver/platform/resource/ResourcesTest.java:14]: Nope, why did you do this?

File: src/security/security-tests/src/test/java/org/geoserver/security/password/URLMasterPasswordProviderTest.java [hunk 15]
```diff
@@ -46,4 +46,33 @@ public void testEncryption() throws Exception {
         char[] passwd = mpp.doGetMasterPassword();
         assertTrue(Arrays.equals("geoserver".toCharArray(), passwd));
     }
+
+    @Test
+    public void testSecuredAccess() throws Exception {
```
Comment (aaime) [inline src/security/security-tests/src/test/java/org/geoserver/security/password/URLMasterPasswordProviderTest.java:51]: This seems like an unrelated change, or not? Please elaborate.

**Review by aaime — CHANGES_REQUESTED** [review]

**Comment by aaime** [comment]
Also, there is no documentation, should be added.

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/RemoteRequestInputProvider.java (hunk 1 above)
Comment (imranrajjad) [reply src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/RemoteRequestInputProvider.java:64]: dont think so

File: src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ExecuteTest.java (hunk 4 above)
Comment (imranrajjad) [reply src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ExecuteTest.java:791]: this error comes from Geotools

File: src/main/src/main/java/org/geoserver/security/urlchecker/GeoserverURLConfigService.java (hunk 10 above)
Comment (imranrajjad) [reply src/main/src/main/java/org/geoserver/security/urlchecker/GeoserverURLConfigService.java:83]: will add

File: src/platform/src/test/java/org/geoserver/platform/resource/ResourcesTest.java (hunk 14 above)
Comment (imranrajjad) [reply src/platform/src/test/java/org/geoserver/platform/resource/ResourcesTest.java:14]: will remove

File: src/security/security-tests/src/test/java/org/geoserver/security/password/URLMasterPasswordProviderTest.java (hunk 15 above)
Comment (imranrajjad) [reply src/security/security-tests/src/test/java/org/geoserver/security/password/URLMasterPasswordProviderTest.java:51]: removed

File: src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ExecuteTest.java (hunk 4 above)
Comment (aaime) [reply src/extension/wps/wps-core/src/test/java/org/geoserver/wps/ExecuteTest.java:790]: Then fix it in GeoTools, it's still part of the HTTP authorization work no?

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/RemoteRequestInputProvider.java (hunk 1 above)
Comment (zvrablik) [reply src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/RemoteRequestInputProvider.java:64]: in the 05-007r7_Web_Processing_Service_WPS_v1.0.0.pdf page 31, there is activity diagram with FTP 
So I think the FTP should be covered.
I don't understand why there is the if statement, I would test all destinations of all protocols. I think all has to be whitelisted including file, http, ftp or whatever new there will be available otherwise it is security hole.
@aaime  Is there something I don't see?
